rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,270 @@
+co-authors: Robert Hensing
+pre-RFC reviewers: Thomas Bereknyei, John Ericson, Alex Ameen
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: https://github.com/NixOS/nixpkgs/pull/211832
+---
+
+# Summary
+[summary]: #summary
+
+Auto-generate trivial top-level attribute definitions in `pkgs/top-level/all-packages.nix` from a directory structure that matches the attribute name.
+This makes it much easier to contribute new packages, since there's no more guessing needed as to where the package should go, both in the ad-hoc directory categories and in `pkgs/top-level/all-packages.nix`.
+
+# Motivation
+[motivation]: #motivation
+
+- It is not obvious to package contributors where to add files or which ones to edit. These are very common questions:
+  - Which directory should my package definition go in?
+  - What are all the categories and do they matter?
+  - What if the package has multiple matching categories?
+  - Why can't I build my package after adding the package file?
+  - Where in all-packages.nix should my package go?
+- Figuring out where an attribute is defined is a bit tricky:
+  - First one has to find the definition of it in all-packages.nix to see what file it refers to
+    - On GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
+  - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
+    - It also slows down or even locks up editors due to the file size
+  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
+- `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages
+
+# Detailed design
+[design]: #detailed-design
+
+This RFC consists of two parts, each of which is implemented with a PR to Nixpkgs.
+
+## PR 1: The unit directory standard
+
+This part establishes the new _unit directory standard_ in Nixpkgs.
+This standard is internal to Nixpkgs and not exposed as public interface.
+This standard must be documented in the Nixpkgs manual.
+
+### File structure
+
+Create the initially-empty directory `pkgs/unit`, called _unit base directory_, in Nixpkgs.
+
+Check the following using CI:
+- The unit base directory must only contain subdirectories of the form `pkgs/unit/${shard}/${name}`, called _unit directories_.
+- `name` is a string only consisting of the ASCII characters a-z, A-Z, 0-9, `-` or `_`.
+- `shard` is the lowercased first two letters of `name`, expressed in Nix: `shard = toLower (substring 0 2 name)`.
+- Each unit directory must contain a `package.nix` file and may contain arbitrary other files.
```

Comment:
A proposal relating to the changes [to remove custom arguments]() and [not enforcing the standard]():

Both the current directory structure and `all-packages.nix` suck, but as @piegamesde also already hinted at, we could just have the unit directory structure without the extra semantic checks, only having these highlighted checks.

While the extra semantics are useful and there for a reason, a lot of attributes don't satisfy them (non-packages, package sets, custom `callPackage` arguments, other `callPackage`s, etc.). According to the current RFC this would not allow all of those attributes to be moved to `pkgs/unit`.

So how about having a _separate_ directory, call it something like `pkgs/sem-less-unit` or `pkgs/manual-unit`. This directory is the same as `pkgs/unit`, but it doesn't get auto-called, you still need to manually edit `all-packages.nix`, but in turn it doesn't come with the restrictions of `pkgs/unit`, allowing all top-level attributes to be migrated to it.

The one semantic restriction we should still want to have though is the one preventing unit directories from accessing files outside themselves, because afaik all such problems are fixable in a better way.

Reply:
> So how about having a _separate_ directory, call it something like `pkgs/sem-less-unit` or `pkgs/manual-unit`. This directory is the same as `pkgs/unit`,

That's over-engineered and incompatible with our goal of having a single place to put a package.